subsys/bluetooth/host/hci_core.c

```diff
@@ -7123,6 +7291,339 @@ bool bt_addr_le_is_bonded(uint8_t id, const bt_addr_le_t *addr)
+		return -EINVAL;
+	} else if (atomic_test_bit(adv->flags, BT_ADV_SCANNABLE)) {
+		return -EINVAL;
+	} else if (atomic_test_bit(adv->flags, BT_ADV_CONNECTABLE)) {
+		return -EINVAL;
+	} else if (atomic_test_bit(adv->flags, BT_ADV_LEGACY)) {
+		return -EINVAL;
+	}
+
+	if (param->interval_min < 0x0006 ||
+	    param->interval_max < 0x0006 ||
+	    param->interval_min > param->interval_max) {
+		return -EINVAL;
+	}
+
+	/* TODO: What if you change the ext_adv_params to e.g. connectable after
+	 * setting the per_adv_params?
+	 */
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_PER_ADV_PARAM, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = adv->handle;
+	cp->min_interval = param->interval_min;
+	cp->max_interval = param->interval_max;
+
+	if (param->options & BT_LE_ADV_OPT_USE_TX_POWER) {
+		cp->props |= BT_HCI_LE_ADV_PROP_TX_POWER;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_PER_ADV_PARAM, buf, NULL);
+	if (err) {
+		return err;
+	}
+
+	atomic_set_bit(adv->flags, BT_PER_ADV_PARAMS_SET);
+
+	return 0;
+}
+
+int bt_le_per_adv_set_data(const struct bt_le_ext_adv *adv, const uint8_t *data,
+			   uint8_t data_len)
+{
+	struct bt_hci_cp_le_set_per_adv_data *cp;
+	struct net_buf *buf;
+	int err;
+
+	if (!atomic_test_bit(adv->flags, BT_PER_ADV_PARAMS_SET)) {
+		return -EINVAL;
+	}
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_PER_ADV_DATA, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = adv->handle;
+	/* TODO: If data is longer than what the controller can manage,
+	 * split the data. Read size from controller on boot.
+	 */
+	cp->op = BT_HCI_LE_EXT_ADV_OP_COMPLETE_DATA;
+	cp->len = data_len;
+	memcpy(cp->data, data, data_len);
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_PER_ADV_DATA, buf, NULL);
+	if (err) {
+		return err;
+	}
+	return 0;
+}
+
+int bt_le_per_adv_enable(struct bt_le_ext_adv *adv, bool enable)
+{
+	struct bt_hci_cp_le_set_per_adv_enable *cp;
+	struct net_buf *buf;
+	int err;
+
+	/* TODO: We could setup some default ext adv params if not already set*/
+	if (!atomic_test_bit(adv->flags, BT_PER_ADV_PARAMS_SET)) {
+		return -EINVAL;
+	}
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_PER_ADV_ENABLE, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = adv->handle;
+	cp->enable = enable ? 1 : 0;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_PER_ADV_ENABLE, buf, NULL);
+	if (err) {
+		return err;
+	}
+
+	if (enable) {
+		atomic_set_bit(adv->flags, BT_PER_ADV_ENABLED);
+	} else {
+		atomic_clear_bit(adv->flags, BT_PER_ADV_ENABLED);
+	}
+	return 0;
+}
+
+#if defined(CONFIG_BT_PER_ADV_SYNC)
+
+uint8_t bt_le_per_adv_sync_get_index(struct bt_le_per_adv_sync *per_adv_sync)
+{
+	uint32_t index = per_adv_sync - per_adv_sync_pool;
```

Comment:
I actually want to change this to a uintptr_t. 

The reasoning behind using such a large type for this, is that if the pointer is indeed out of the index, then we may actually risk that the overflow when using a uint8_t makes it a valid index. 

Example if `per_adv_sync_pool` is `0x12340000` and `per_adv_sync` is `0x12340100` then the result will be be `0` rather than `256`, where `0` is a valid index but `256` certainly is not. 

Of course this requires that the application are messing with some bad stuff, but I'd rather actually catch it here than expecting the application to behave. 

I can change it back to a 8-bit value if you rather want that (it's an application error anyhow), but otherwise I'd suggest that we consider the same for the ext adv index lookup.

Reply:
Ok. let's use uintptr_t, and update the ext_adv _index lookup as well.